dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java

```diff
@@ -273,7 +273,15 @@ private void extractEventsFromModel() {
                     e = e.getSuccessor();
                     continue;
                 }
-                if (eventFilter.apply(e)) {
+                // In the new model we extract visible events, Locals and Asserts to show them in the
+                // witness, and extract also CondJumps for tracking internal dependencies. For consistency
+                // of the view of events which infects the derived relation computation by CAAT, we need 
+                // to extract the same events here in the old model.
+                if (eventFilter.apply(e)
+                    || e instanceof Local
+                    || e instanceof Assert
+                    || e instanceof CondJump
+                ) {
```

Comment:
This should be in the old model. The two models should extract the same events, because the computation of derived relations and the construction of the `EdgeModels` in witness should be based on the same events view.

Reply:
I don't understand. The old model is only used for CAAT/Refinement, no? And CAAT does not understand these events. This is bound to break the Refinement Solver and is likely to be the culprit for your CI failures.
I don't think you should try to bring the old model and the new model under a common interface.
If you want Refinement to generate executions of the new shape, just make a simple adaption: let it create both the old and the new execution model in each iteration: the old one is used for refinement, the new one as a final witness. In fact, you might even be able to only generate the new model in the final iteration where it finds a true counterexample. Then its not even wasteful to just create both types of execution models.

Reply:
Let me ask a question: if I construct and populate a `SimpleGraph` for `DirectDataDependency`, this relation involves invisible events and is `PredicateHierarchy` able to correctly compute the derived relations based on it? And for the CI failures: If introducing the invisibles causes them, I would expect more failures rather than less than 5 each time.

Reply:
The code of the `CAATSolver` and the `PredicateHierarchy` do not care at all about visible/invisible events. What cares about those is the `WmmSolver` and its custom graphs like `ProgramOrderGraph`. The `ProgramOrderGraph` assumes that all events in the `ExecutionModel` are visible one's (even if they are not) and as such considers them to be related by `po`, i.e., you can end up with invisible events in the program order. For monotonic memory models, this might be a non-issue because the extra events will not contribute to anything really, but for non-monotonic one's you can get problems.

In short: you can have invisible events in your own graphs and that is totally fine, but you cannot have those events in the graphs that are used for the reasoning in the `RefinementSolver`. However, your current approach adds those, potentially breaking the solver.

Reply:
OK I think I got what you mean. But one problem about my question is: we need to initialize `PredicateHierarchy` to a `EventDomain`, and so far this domain can be constructed from the old model only. This is the reason why I added invisibles. To fix it, what I should do is to let the `EventDomain` be able to be constructed also from the new model instead of adding new events into the old one, right?

Reply:
OK. I will implement this new domain, undo the change to the old model and `WMMSolver`, and see if this can solve the failures.